ssz-rs/src/de.rs

```diff
@@ -93,26 +102,41 @@ where
         return Ok(vec![])
     }
 
-    let data_pointer = u32::deserialize(&encoding[..BYTES_PER_LENGTH_OFFSET])?;
-    let data_pointer = data_pointer as usize;
-    if encoding.len() < data_pointer {
+    let offsets_len = encoding.get(..BYTES_PER_LENGTH_OFFSET).ok_or({
+        DeserializeError::ExpectedFurtherInput {
+            provided: encoding.len(),
+            expected: BYTES_PER_LENGTH_OFFSET,
+        }
+    })?;
+    let offsets_len = u32::deserialize(offsets_len)?;
+    let offsets_len = offsets_len as usize;
+    if encoding.len() < offsets_len {
         return Err(DeserializeError::ExpectedFurtherInput {
             provided: encoding.len(),
-            expected: data_pointer,
+            expected: offsets_len,
         })
     }
+    if offsets_len % BYTES_PER_LENGTH_OFFSET != 0 {
+        return Err(DeserializeError::InvalidOffsetsLength(offsets_len))
+    }
 
-    let offsets = &mut encoding[..data_pointer]
+    let offsets = &mut encoding[..offsets_len]
         .chunks_exact(BYTES_PER_LENGTH_OFFSET)
         .map(|chunk| u32::deserialize(chunk).map(|offset| offset as usize))
         .collect::<Result<Vec<usize>, DeserializeError>>()?;
     offsets.push(encoding.len());
 
-    let element_count = data_pointer / BYTES_PER_LENGTH_OFFSET;
+    let element_count = offsets_len / BYTES_PER_LENGTH_OFFSET;
     let mut result = Vec::with_capacity(element_count);
     for span in offsets.windows(2) {
+        // SAFETY: index is safe because span is a pair; qed
         let start = span[0];
         let end = span[1];
+        if start > end {
+            return Err(DeserializeError::OffsetNotIncreasing { start, end })
+        }
+
+        // SAFETY: index is safe because start <= end; qed
```

Comment:
and `start` can be at *least* 0 and `end` can be at *most* the length of the encoding